// tedious/lib/token/returnvalue-token-parser.js
var _regeneratorRuntime = require('babel-runtime/regenerator').default;

// Generated by CoffeeScript 1.10.0
var metadataParse, valueParse;

metadataParse = require('../metadata-parser');

valueParse = require('../value-parser');

module.exports = _regeneratorRuntime.mark(function callee$0$0(parser, colMetadata, options) {
  var metadata, paramName, paramOrdinal, status, token, value;
  return _regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return parser.readUInt16LE();

      case 2:
        paramOrdinal = context$1$0.sent;
        return context$1$0.delegateYield(parser.readBVarChar(), 't0', 4);

      case 4:
        paramName = context$1$0.t0;
        context$1$0.next = 7;
        return parser.readUInt8();

      case 7:
        status = context$1$0.sent;
        return context$1$0.delegateYield(metadataParse(parser, options), 't1', 9);

      case 9:
        metadata = context$1$0.t1;
        return context$1$0.delegateYield(valueParse(parser, metadata, options), 't2', 11);

      case 11:
        value = context$1$0.t2;

        if (paramName.charAt(0) === '@') {
          paramName = paramName.slice(1);
        }
        return context$1$0.abrupt('return', token = {
          name: 'RETURNVALUE',
          event: 'returnValue',
          paramOrdinal: paramOrdinal,
          paramName: paramName,
          metadata: metadata,
          value: value
        });

      case 14:
      case 'end':
        return context$1$0.stop();
    }
  }, callee$0$0, this);
});
